bloom.go

```diff
@@ -0,0 +1,89 @@
+// Package bloom implements a Bloom filter data structure.
+package bloom
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"log"
+	"math/bits"
+)
+
+// Filter represents a Bloom filter. Filter satisfies the encoding.BinaryMarshaler and BinaryUnmarshaler interfaces.
+// The zero value represents an empty filter of size 0 that uses 0 hash values.
+// Current implementations of Insert and Contains require len(f) <= 2^16 bits and k <= 16.
+type Filter struct {
+	f []byte
+	k int
+}
+
+// bit returns the filter's nth bit.
+func (f *Filter) bit(n int) int {
+	b, i := n/8, n%8
+	return int(f.f[b]>>uint(i)) & 1
+}
+
+// setBit sets the filter's nth bit to 1.
+func (f *Filter) setBit(n int) {
+	b, i := n/8, n%8
+	f.f[b] |= 1 << uint(i)
+}
+
+// hashBits returns the base 2 logarithm of the size of the filter in bits.
+// This is the number of hash bits required to index any bit in the filter.
+// TODO: Handle filter size not a power of 2
+func (f *Filter) hashBits() int {
+	if bits.OnesCount(uint(len(f.f))) != 1 {
+		log.Fatal("hashBits: filter size not a power of 2")
```

Comment:
Libraries should generally never call log.Fatal. Either return an error, or panic if there is a bug. (Or, if we check for this in New & UnmarshalBinary, then no need to check again here imo.)